### PR TITLE
Add inital support for assigning groups from Okta response.

### LINKED ISSUE #42: Allow pulling in Okta group info and map to teamcity group
Hello,

Is there a way to pull in group information from okta and map it to teamcity groups? If not, can this be filed as an enhancement request?

### PR DESCRIPTION
**Pull request by fatmcgav**
This is an issue commit that adds support for assigning TeamCity groups
based on the SAML response from OKTA.

New SAML response was generated using Samling, with 2 groups assigned.

Fixes #42

### CODE REVIEW

**Comment by fatmcgav** [comment]
@morincer I'd be interested in getting some early feedback on this PR.

Do the changes look sane to you? Any issues you can foresee or additional functionality you'd like to see added?

**Comment by morincer** [comment]
Thanks for the PR - I'll review it today and get back to you.

**Comment by fatmcgav** [comment]
Cool, cheers.

I'm working on the UI component currently, so should have that ready later this evening or tomorrow morning BST.

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java [hunk 1]
```diff
@@ -139,6 +144,22 @@ public HttpAuthenticationResult processAuthenticationRequest(@NotNull HttpServle
 
                             if (user == null) {
                                 LOG.warn(String.format("New user %s was not created due to unknown reason", username));
+                            } else {
+                                String groups = getAttribute(auth, settings.getGroupsAttributeMapping());
+                                LOG.debug(String.format("SAML Groups = '%s'", groups));
+
+                                // Look for matching userGroup
+                                String[] assignedGroups = groups.split(",");
```
Comment (morincer) [inline saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java:152]: Error here - getAttribute joins collections with ", " (comma with space) not ",". Anyway, I think it's better to add some additional processing to the list - trim and filter empty records. Something like
```
Arrays.stream(groups.split(", "))
                                        .map(String::trim)
                                        .filter(StringUtil::isNotEmpty)
                                        .forEach(...);
```

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java [hunk 2]
```diff
@@ -139,6 +144,22 @@ public HttpAuthenticationResult processAuthenticationRequest(@NotNull HttpServle
...
+                                for (String group: assignedGroups) {
+                                    SUserGroup matchingGroup = userGroups.findUserGroupByName(group);
```
Comment (morincer) [inline saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java:154]: Meanwhile this works, I would rather do it bit differently - I would fetch the list of all user groups names into a Hashset<STring> before the loop and just use it as a lookup. It has another advantage - we could additionaly lowercase/uppercase groups in the SAML attribute and group names from Teamcity and make the mapping case-insensitive.

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java [hunk 3]
```diff
@@ -139,6 +144,22 @@ public HttpAuthenticationResult processAuthenticationRequest(@NotNull HttpServle
 
                             if (user == null) {
                                 LOG.warn(String.format("New user %s was not created due to unknown reason", username));
+                            } else {
```
Comment (morincer) [inline saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java:147]: Please note, that placing the logic under this conditional branch only enables it for newly created users. It won't allow changing groups set for existing users.

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlPluginConfiguration.java [hunk 4]
```diff
@@ -23,8 +24,8 @@
 public class SamlPluginConfiguration {
 
     @Bean
-    SamlAuthenticationScheme samlAuthenticationScheme(LoginConfiguration loginConfiguration, SamlPluginSettingsStorage samlPluginSettingsStorage, UserModel userModel, RootUrlHolder rootUrlHolder) {
-        SamlAuthenticationScheme samlAuthenticationScheme = new SamlAuthenticationScheme(rootUrlHolder, samlPluginSettingsStorage, userModel, loginConfiguration);
+    SamlAuthenticationScheme samlAuthenticationScheme(LoginConfiguration loginConfiguration, SamlPluginSettingsStorage samlPluginSettingsStorage, UserModel userModel, UserGroupManager userGroups, RootUrlHolder rootUrlHolder) {
```
Comment (morincer) [inline saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlPluginConfiguration.java:27]: Please use consistent attributes naming userGroupsManager not userGroups

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java (hunk 3 above)
Comment (fatmcgav) [reply saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java]: I've moved this logic as part of [`5a2d82e` (#52)](https://github.com/morincer/teamcity-plugin-saml/pull/52/commits/5a2d82e3e9167f316bb5d31cf2a24c11c706b4dc) in order to put it behind a boolean setting. 

Will need to add logic to remove groups aswell..

**Comment by fatmcgav** [comment]
@morincer So I've just added some additional commits which I think fulfil your comments.

I'm still struggling to test the Ui changes though. I've followed the steps outlined in the README, but for whatever reason the new fields added in 5a2d82e3e9167f316bb5d31cf2a24c11c706b4dc don't actually render.
Are you able to provide any pointers?

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java (hunk 2 above)
Comment (fatmcgav) [reply saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java]: Addressed in [`732591b` (#52)](https://github.com/morincer/teamcity-plugin-saml/pull/52/commits/732591b0503ddf1447ad56eb81d5497aad25b4b0)

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java (hunk 1 above)
Comment (fatmcgav) [reply saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java]: Addressed in [`732591b` (#52)](https://github.com/morincer/teamcity-plugin-saml/pull/52/commits/732591b0503ddf1447ad56eb81d5497aad25b4b0)

**Comment by morincer** [comment]
@fatmcgav , well, I just usually execute "mvn package tc-sdk:reload" from the repository root and then reload plugin in admin UI and it does the trick.
I'll check your code.

**Comment by fatmcgav** [comment]
> @fatmcgav , well, I just usually execute "mvn package tc-sdk:reload" from the repository root and then reload plugin in admin UI and it does the trick.
> I'll check your code.

So I managed to get the UI to reload by nuking the `servers` dir and then re-starting stuff...

One issue I have just come across is that Okta doesn't have a friendly "Human" name on the response. Instead, it looks like it can pass explicit fields for `user.firstName` and `user.lastName`. 

What do you think about adding a "Composite" attribute type which can be used to join a comma separated list of SAML attributes?

**Comment by morincer** [comment]
> One issue I have just come across is that Okta doesn't have a friendly "Human" name on the response. Instead, it looks like it can pass explicit fields for `user.firstName` and `user.lastName`.
> 
> What do you think about adding a "Composite" attribute type which can be used to join a comma separated list of SAML attributes?

Manipulating SAML attributes is not what exactly the plugin is for. You can achieve the same on IdP level - Okta allows to use expressions when defining a custom attribute - 

image

**Comment by fatmcgav** [comment]
> > One issue I have just come across is that Okta doesn't have a friendly "Human" name on the response. Instead, it looks like it can pass explicit fields for `user.firstName` and `user.lastName`.
> > What do you think about adding a "Composite" attribute type which can be used to join a comma separated list of SAML attributes?
> 
> Manipulating SAML attributes is not what exactly the plugin is for. You can achieve the same on IdP level - Okta allows to use expressions when defining a custom attribute -
> 
> image

Yeh, after I posted the comment, I managed to find an example in the Okta KB that set the value to `user.fullName`. This value doesn't render in the drop-down, which is what made me think Okta didn't support it...

For reference, my Okta config looks like:
image

**Comment by fatmcgav** [comment]
@morincer So I've just pushed another commit which I think rounds the code out fully.

In testing against a Dev okta instance, I'm able to create a new user in TC with groups assigned, and add and remove TC group assignments based on Okta group membership.

LMK what you think :)

**Comment by fatmcgav** [comment]
I've also just fixed the merge conflict :)

**Comment by fatmcgav** [comment]
A quick example of the login flow:
TC-Login-SSO

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java [hunk 5]
```diff
@@ -150,6 +154,16 @@ public HttpAuthenticationResult processAuthenticationRequest(@NotNull HttpServle
                 return sendUnauthorizedRequest(request, response, String.format("SAML request NOT authenticated for user id %s: user with such username or %s property value not found", username, SamlPluginConstants.ID_USER_PROPERTY_KEY));
             }
 
+            if (settings.isAssignGroups()) {
+                String samlGroups = getAttribute(auth, settings.getGroupsAttributeMapping());
+                LOG.debug(String.format("SAML Groups = '%s'", samlGroups));
+
+                // Process the SAML groups assigned to this user
+                if (!samlGroups.isEmpty()) {
```
Comment (morincer) [inline saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java:162]: Here you've lost a use case when somebody wants to completely remove user from all the groups - i.e. if I set "remove unassigned groups = True" I would expect that the groups will be cleared for the user, not just skipped. I would rather remove this condition and let the processGroups logic handle the rest (don't forget the null check of samlGroups then)

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java [hunk 6]
```diff
@@ -196,6 +210,68 @@ private boolean matchPostfixes(String username, String allowedPostfixes) {
...
+                .filter(StringUtil::isNotEmpty)
+                .collect(Collectors.toList());
+        LOG.debug(String.format("Users assigned groups from SAML response: '%s'", usersAssignedGroups));
+
+        // What groups to add and what groups to remove
+        List<String> groupsToAdd = new ArrayList<>(CollectionUtils.subtract(usersAssignedGroups, usersCurrentGroups));
+
+        // Add any new groups
+        groupsToAdd.forEach(addGroup -> {
+           SUserGroup tcGroup = teamcityGroups.stream()
```
Comment (morincer) [inline saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java:239]: Not the best way to lookup things - you literally iterating the collection over and over again (and do it twice - here and below). Instead just make teamcityGroups a map (Collectors.toMap(...) with lowercased group key) and use .containsKey

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java [hunk 7]
```diff
@@ -196,6 +210,68 @@ private boolean matchPostfixes(String username, String allowedPostfixes) {
...
+           tcGroup.addUser(user);
+        });
+
+        // Optionally remove groups that are no longer assigned in SAML response.
+        if (removeUnassignedGroups) {
+            List<String> groupsToRemove = new ArrayList<>(CollectionUtils.subtract(usersCurrentGroups, usersAssignedGroups));
+
+            // Remove any groups that are no longer mapped
+            groupsToRemove.forEach(removeGroup -> {
+                SUserGroup tcGroup = teamcityGroups.stream()
```
Comment (morincer) [inline saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java:259]: See above

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java (hunk 5 above)
Comment (fatmcgav) [reply saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java:162]: Ah, good point... Will update.

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java (hunk 6 above)
Comment (fatmcgav) [reply saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java:239]: Heh, this is my first time writing Java, so thanks for the pointers.

I'll rework.

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java (hunk 5 above)
Comment (fatmcgav) [reply saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java]: Addressed in [`8878dcc` (#52)](https://github.com/morincer/teamcity-plugin-saml/pull/52/commits/8878dccd3ef2513420554cdeb65d19de6d711d59)

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java (hunk 6 above)
Comment (fatmcgav) [reply saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java]: I think I've addressed in [`8878dcc` (#52)](https://github.com/morincer/teamcity-plugin-saml/pull/52/commits/8878dccd3ef2513420554cdeb65d19de6d711d59)

Tests still all pass atleast :)

File: saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java (hunk 7 above)
Comment (fatmcgav) [reply saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java]: [`8878dcc` (#52)](https://github.com/morincer/teamcity-plugin-saml/pull/52/commits/8878dccd3ef2513420554cdeb65d19de6d711d59)

**Comment by fatmcgav** [comment]
Right, I think this is ready for another look @morincer :)

**Comment by fatmcgav** [comment]
@morincer thanks for merging.

Is it possible to get a new release cut with this included?

**Comment by morincer** [comment]
Just released 1.3 with this change (just added small null change).

**Comment by fatmcgav** [comment]
Great, thanks for that @morincer

**Comment by anherrera** [comment]
This change is great! Allows us to use Okta with Teamcity!

I ran into two edge cases and I solved one of them but not the other: 

1. Okta group has spaces... Replace the `\w+` regex with just `.*`
2. Okta group has special characters, such as `&` - how do I set the group key to account for this? If I have `&` in the group key at all TeamCity breaks and won't add roles to it. Woops. I suppose we could create another group name in Okta without `&` in it but I was hoping there was another way.

**Comment by morincer** [comment]
Hi @anherrera 
Are you able to show me a portion of the SAML request with failure group names so I run a couple of tests for it?
In general, I guess a good approach would be to replace non-valid group symbols with _ or something like that.

**Pull request merged**